### PR TITLE
[wip] recognize VMs in Ambient

### PR DESCRIPTION
**Pull request by adiprerepa**

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -1226,6 +1226,40 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
...
+		Status: v1.PodStatus{
+			PodIP: si.Endpoint.Address,
+			Conditions: []v1.PodCondition{{
+				Type:   v1.PodReady,
+				Status: v1.ConditionTrue,
+			}},
+		},
+	}
+	// treat workloads on ambient as HBONE native
+	ambientMockPod.Labels[model.TunnelLabel] = model.TunnelHTTP
```
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1254]: I don't think this should be intrinsic. The user should stick this on their workload entry if it's running in that mode.

**Comment by linsun** [comment]
Nice to see some progress on this - Would it be reasonable to have a design doc on this before this is implemented/merged? 

cc @howardjohn @shashankram

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 1 above)
Comment (adiprerepa) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:1254]: okay, will delete. Just going to let it be passed through si.Endpoint.Labels

**Comment by adiprerepa** [comment]
> Nice to see some progress on this - Would it be reasonable to have a design doc on this before this is implemented/merged?
@linsun not sure it really makes sense, it's a simple and intuitive enough change that i'm not really sure what would be expressed in a design doc. If you all disagree however, I'm happy to make one.

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -1226,6 +1226,40 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
...
+	for k, v := range si.PortMap {
+		ambientMockPod.Spec.Containers[0].Ports = append(ambientMockPod.Spec.Containers[0].Ports, v1.ContainerPort{
+			Name:          k,
+			ContainerPort: int32(v),
+		})
+	}
+	c.ambientIndex.handlePods([]*v1.Pod{ambientMockPod})
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1261]: ambientIndex can be nil

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 3]
```diff
@@ -1226,6 +1226,40 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
 		c.workloadInstancesIndex.Insert(si)
 	}
 
+	// handle WorkloadInstance in ambient by creating a mock pod that models the VM workload.
+	ambientMockPod := &v1.Pod{
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1230]: Suggest making this handling in ambientindex.go, maybe add a new function AmbientIndex.handleWorkloadInstance

**Comment by linsun** [comment]
> > Nice to see some progress on this - Would it be reasonable to have a design doc on this before this is implemented/merged?
> 
> @linsun not sure it really makes sense, it's a simple and intuitive enough change that i'm not really sure what would be expressed in a design doc. If you all disagree however, I'm happy to make one.

Thanks @adiprerepa! I added this as an agenda item for today's ambient meeting, would like to learn more.

**Comment by stevenctl** [comment]
For the simple case of "include WorkloadEntry endpoints in WDS" I don't know if that needs to be blocked by design. 
WRT making the zTunnel itself run properly outside a "shared node proxy" context, we have some issues to discuss.

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 4]
```diff
@@ -1226,6 +1226,40 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
 		c.workloadInstancesIndex.Insert(si)
 	}
 
+	// handle WorkloadInstance in ambient by creating a mock pod that models the VM workload.
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1229]: I think we need the same logic we use for the rest of Istio? We need ServiceEntries as well, and ServiceEntry selecting Pod, Service selecting WorkloadEntry, etc

Comment (linsun) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:1229]: +1

**Comment by imroc** [comment]
> WRT making the zTunnel itself run properly outside a "shared node proxy" context, we have some issues to discuss

@stevenctl I think this design make it possible to run ZT properly outside a "shared node proxy" context: https://docs.google.com/document/d/135gWueRb5KW2vc0RerE2vGKBoCFxbqoLVBuTGJBcxO4/edit#

Related PR: https://github.com/istio/ztunnel/pull/452

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 2 above)
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1261]: This assumes its an "add". Could be an update or a delete.

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 5]
```diff
@@ -1226,6 +1226,40 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
...
+			NodeName:           si.Endpoint.NodeName,
+			Containers: []v1.Container{
+				{
+					Ports: []v1.ContainerPort{},
+				},
+			},
+		},
+		Status: v1.PodStatus{
+			PodIP: si.Endpoint.Address,
+			Conditions: []v1.PodCondition{{
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1247]: I think the WI can be "not ready"

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 6]
```diff
@@ -1226,6 +1226,40 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
...
+	// handle WorkloadInstance in ambient by creating a mock pod that models the VM workload.
+	ambientMockPod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      si.Name,
+			Namespace: si.Namespace,
+			Labels:    si.Endpoint.Labels,
+		},
+		Spec: v1.PodSpec{
+			ServiceAccountName: si.Endpoint.ServiceAccount,
+			NodeName:           si.Endpoint.NodeName,
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:1238]: this is always empty on a WorkloadEntry fwiw

**Review by howardjohn — COMMENTED** [review]
I don't feel great about pretending to be a Pod

**Comment by howardjohn** [comment]
Thanks for kicking this off! I think this has been superseded with some other PRs based on this work

**Pull request closed without merging**